### PR TITLE
Wallpapers

### PR DESCRIPTION
**Pull request by gordonbrander**
- [x] Theme dashboard
- [x] Theme navigation
- [x] Theme awesomebar
- [x] Update dashboard wallpaper and navigation on swatch click
- [x] Configure real themes
- [x] Retain chosen theme after refresh

### CODE REVIEW

**Comment by gordonbrander** [comment]
Fixes https://github.com/mozilla/browser.html/issues/214

**Comment by gordonbrander** [comment]
@Gozala r?

File: css/dashboard.css [hunk 1]
```diff
@@ -1,14 +1,24 @@
 .dashboard {
+  background-position: center center;
+  background-size: cover;
+  align-items: center;
+  flex: 1 0 auto;
+  flex-flow: column nowrap;
+  color: #555;
+  justify-items: center;
```
Comment (paulrouget) [inline css/dashboard.css]: This property doesn't exist.

File: css/awesomebar.css [hunk 2]
```diff
@@ -34,29 +34,29 @@
   line-height: 38px;
   vertical-align: center;
```
Comment (paulrouget) [inline css/awesomebar.css]: s/center/middle (… I wrote that :/ )

File: css/dashboard.css [hunk 3]
```diff
@@ -50,8 +59,26 @@
   opacity: 1;
 }
 
+.wallpaper-swatches {
+  flex-flow: row nowrap;
```
Comment (paulrouget) [inline css/dashboard.css]: I would prefer if we could avoid shorthands for flex properties. I'm not familiar enough with them yet.

File: src/browser/dashboard/actions.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    );
+  }
+
+  // Exports:
+
+  exports.getDashboardThemePatch = getDashboardThemePatch;
+  exports.getWallpaperSwatches = getWallpaperSwatches;
+  exports.readDashboardTheme = readDashboardTheme;
+  exports.readDashboardNavigationTheme = readDashboardNavigationTheme;
+  exports.init = init;
```
Comment (paulrouget) [inline src/browser/dashboard/actions.js]: All the methods use "dashboard" in their names. Do the same thing with `init`.

File: src/browser/dashboard/actions.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+define((require, exports, module) => {
+  'use strict';
+
+  const {expandCustomTheme} = require('../theme.js');
+
+  // Like `array.reduce` but for objects.
```
Comment (paulrouget) [inline src/browser/dashboard/actions.js]: `Object.keys(object).reduce()`

File: src/browser/dashboard/actions.js [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  const reducekv = (object, step, value) => {
+    for (let key in object) {
+      value = step(value, key, object[key]);
```
Comment (paulrouget) [inline src/browser/dashboard/actions.js]: What about using `Object.keys(object).reduce()`?

File: src/browser/actions.js [hunk 7]
```diff
@@ -81,7 +82,7 @@ define((require, exports, module) => {
     // once #239 is resolved.
     input: {value: '', isFocused: false},
     tabStrip: {isActive: false},
-    dashboard: {items: dashboardItems},
+    dashboard: initDashboard({items: dashboardItems}),
```
Comment (paulrouget) [inline src/browser/actions.js]: What about:
```
dashboard: {
  items: dashboardItems,
  theme: getDashboardThemePatch('default')
}
```
and forget about the `init` function?

File: src/browser/browser.js [hunk 8]
```diff
@@ -165,6 +167,7 @@ define((require, exports, module) => {
     const rfaCursor = immutableState.cursor('rfa');
 
     const dashboard = immutableState.get('dashboard');
+    const dashboardCursor = immutableState.cursor('dashboard');
```
Comment (paulrouget) [inline src/browser/browser.js]: Maybe you want to use the cursor everywhere, and kill `dashboard`.

File: src/browser/dashboard.js [hunk 9]
```diff
@@ -7,12 +7,27 @@ define((require, exports, module) => {
...
-    ({key, uri, image, title}, {onOpen}) =>
+  const List = (Item, a2b) => Component('List', (options, handlers) =>
+    DOM.div(options, options.items.map(options =>
+      Item(a2b(options), handlers))));
+
+  const WallpaperSwatch = ({key, backgroundColor}, {onWallpaperChange}) =>
+    DOM.div({
+      key,
+      className: 'wallpaper-swatch',
+      style: {backgroundColor: backgroundColor},
```
Comment (paulrouget) [inline src/browser/dashboard.js]: `{backgroundColor}`

**Comment by paulrouget** [comment]
You can land this.

File: src/browser/dashboard/actions.js (hunk 4 above)
Comment (gordonbrander) [inline src/browser/dashboard/actions.js]: Was trying to imitate `webviewer/actions.js` which uses short verbs for data model actions. I can change it though. While I'm at it, `createDashboard` is a better name.

File: src/browser/actions.js (hunk 7 above)
Comment (gordonbrander) [inline src/browser/actions.js]: Would rather not. Purpose of function is to be able to add/change data model fields in a central location. Analogous to `webviewer/actions.js` `open({})` function.

File: src/browser/browser.js (hunk 8 above)
Comment (gordonbrander) [inline src/browser/browser.js]: I thought it would be better not to pass cursor unless I need write access.

File: css/dashboard.css (hunk 3 above)
Comment (gordonbrander) [inline css/dashboard.css]: `flex-flow` just combines `flex-direction` and `flex-wrap` https://css-tricks.com/snippets/css/a-guide-to-flexbox/

Comment (paulrouget) [inline css/dashboard.css]: Sure. But let's keep flex box simple and name what we use.

**Pull request merged**